File: ClassicLib/ScanLog/SuspectScanner.py
"""Suspect scanner module for CLASSIC.

This module scans for known crash patterns and suspects including:
- Checking main errors against known patterns
- Scanning call stacks for problematic signatures
- Identifying DLL-related crashes
- Matching against YAML-defined suspect patterns
"""

from typing import TYPE_CHECKING

from ClassicLib.Logger import logger
from ClassicLib.rust.report_rust import ReportFragment
from ClassicLib.ScanLog.scanloginfo import ClassicScanLogsInfo

if TYPE_CHECKING:
    from ClassicLib.ScanLog.scanloginfo import ClassicScanLogsInfo


class SuspectScanner:
    """Provide functionality to scan crash logs for suspect errors using predefined lists of errors and stack
    signatures.

    The `SuspectScanner` class analyzes crash logs and call stack details to identify potential suspect errors.
    It uses predefined lists stored in `ClassicScanLogsInfo` to match against errors and stack traces and report
    findings accordingly.

    Attributes:
        yamldata (ClassicScanLogsInfo): Configuration data containing the suspect error and stack lists used
            for analysis.

    """

    def __init__(self, yamldata: "ClassicScanLogsInfo") -> None:
        """Initialize the suspect scanner.

        Args:
            yamldata: Configuration data containing suspect patterns

        """
        self.yamldata: ClassicScanLogsInfo = yamldata

    def suspect_scan_mainerror(self, crashlog_mainerror: str, max_warn_length: int) -> tuple[ReportFragment, bool]:
        """Scan the crash log for errors listed in a predefined suspect error list.

        Args:
            crashlog_mainerror: The main error output from a crash log to scan for suspect errors.
            max_warn_length: The maximum length for formatting the error name in the report.

        Returns:
            Tuple of (ReportFragment containing findings, bool indicating if suspects found).

        """
        lines = []
        found_suspect = False

        for error_key, signal in self.yamldata.suspects_error_list.items():
            # Skip checking if signal not in crash log
            if signal not in crashlog_mainerror:
                continue

            # Parse error information
            error_severity, error_name = error_key.split(" | ", 1)

            # Format the error name for report
            formatted_error_name = error_name.ljust(max_warn_length, ".")

            # Add the error to the report
            lines.extend((f"- **Checking for {formatted_error_name} SUSPECT FOUND! > Severity : {error_severity}** \n\n", "-----\n"))

            # Update suspect found status
            found_suspect = True

        return ReportFragment.from_lines(lines), found_suspect  # pyright: ignore[reportUnknownArgumentType]

    def suspect_scan_stack(
        self, crashlog_mainerror: str, segment_callstack_intact: str, max_warn_length: int
    ) -> tuple[ReportFragment, bool]:
        """Analyze a crash report and call stack information to identify potential suspect errors.

        Args:
            crashlog_mainerror: The main error extracted from the crash log.
            segment_callstack_intact: The intact segment of the call stack relevant to the analysis.
            max_warn_length: Maximum allowed length for warnings included in the report.

        Returns:
            Tuple of (ReportFragment containing findings, bool indicating if suspects found).

        """
        lines = []
        any_suspect_found = False

        for error_key, signal_list in self.yamldata.suspects_stack_list.items():
            # Parse error information
            error_severity, error_name = error_key.split(" | ", 1)

            # Initialize match status tracking dictionary
            match_status = {
                "has_required_item": False,
                "error_req_found": False,
                "error_opt_found": False,
                "stack_found": False,
            }

            # Process each signal in the list
            should_skip_error = False
            for signal in signal_list:
                # Process the signal and update match_status accordingly
                if self._process_signal(signal, crashlog_mainerror, segment_callstack_intact, match_status):
                    should_skip_error = True
                    break

            # Skip this error if a condition indicates we should
            if should_skip_error:
                continue

            # Determine if we have a match based on the processed signals
            if self._is_suspect_match(match_status):
                # Add the suspect to the report and update the found status
                formatted_error_name = error_name.ljust(max_warn_length, ".")
                lines.extend((f"- **Checking for {formatted_error_name} SUSPECT FOUND! > Severity : {error_severity}** \n\n", "-----\n"))
                any_suspect_found = True

        return ReportFragment.from_lines(lines), any_suspect_found  # pyright: ignore[reportUnknownArgumentType]

    @staticmethod
    def _process_signal(signal: str, crashlog_mainerror: str, segment_callstack_intact: str, match_status: dict[str, bool]) -> bool:
        """Process an individual signal and update match status.

        Returns:
            True if processing should stop (NOT condition met)

        """
        # Debug signal type if not string
        if not isinstance(signal, str):  # pyright: ignore[reportUnnecessaryIsInstance]
            logger.debug(f"_process_signal received non-string signal: {type(signal)} - {signal}")

        # Constants for signal modifiers
        main_error_required = "ME-REQ"
        main_error_optional = "ME-OPT"
        callstack_negative = "NOT"

        if "|" not in signal:
            # Simple case: direct string match in callstack
            if signal in segment_callstack_intact:
                match_status["stack_found"] = True
            return False

        signal_modifier, signal_string = signal.split("|", 1)

        # Process based on signal modifier
        if signal_modifier == main_error_required:
            match_status["has_required_item"] = True
            if signal_string in crashlog_mainerror:
                match_status["error_req_found"] = True
        elif signal_modifier == main_error_optional:
            if signal_string in crashlog_mainerror:
                match_status["error_opt_found"] = True
        elif signal_modifier == callstack_negative:
            # Return True to break out of the loop if NOT condition is met
            return signal_string in segment_callstack_intact
        elif signal_modifier.isdecimal():
            # Check for minimum occurrences
            min_occurrences = int(signal_modifier)
            if segment_callstack_intact.count(signal_string) >= min_occurrences:
                match_status["stack_found"] = True

        return False

    @staticmethod
    def _is_suspect_match(match_status: dict[str, bool]) -> bool:
        """Determine whether a match is classified as a suspect based on the given match status criteria.

        This function evaluates the provided match status dictionary to determine if specific conditions
        indicate a suspect match. Depending on the presence of certain flags, it either confirms the
        criteria are met or checks for alternative flags to ascertain the match status.

        Args:
            match_status (dict[str, bool]): A dictionary containing flags that represent different
                match conditions. Expected keys include:
                - "has_required_item": Indicates the presence of a required item.
                - "error_req_found": Indicates an error condition related to required items.
                - "error_opt_found": Indicates an error condition related to optional items.
                - "stack_found": Indicates whether a stack-related condition was encountered.

        Returns:
            bool: True if the match is determined to be suspect based on the flags in match_status,
            False otherwise.

        """
        if match_status["has_required_item"]:
            return match_status["error_req_found"]
        return match_status["error_opt_found"] or match_status["stack_found"]

    @staticmethod
    def _format_suspect_message(error_name: str, error_severity: str, max_warn_length: int) -> str:
        """Format a warning message for a suspected error condition. The function generates a formatted string
        that includes the name of the error padded to a specified maximum warning length, along with its
        severity level.

        Args:
            error_name (str): The name of the error being checked.
            error_severity (str): The severity level of the error being reported.
            max_warn_length (int): The maximum length to which the error name should be padded.

        Returns:
            str: A formatted string describing the suspected error with its severity.

        """
        formatted_error_name = error_name.ljust(max_warn_length, ".")
        return f"- **Checking for {formatted_error_name} SUSPECT FOUND! > Severity : {error_severity}** \n\n-----\n"

    @staticmethod
    def check_dll_crash(crashlog_mainerror: str) -> ReportFragment:
        """Analyze a crash log to identify if a DLL file is implicated in the crash.

        Args:
            crashlog_mainerror: The main error message extracted from the crash log.

        Returns:
            ReportFragment containing DLL crash notification, or empty fragment.

        """
        crashlog_mainerror_lower = crashlog_mainerror.lower()
        if ".dll" in crashlog_mainerror_lower and "tbbmalloc" not in crashlog_mainerror_lower:
            return ReportFragment.from_lines([
                "* NOTICE : MAIN ERROR REPORTS THAT A DLL FILE WAS INVOLVED IN THIS CRASH! * \n",
                "If that dll file belongs to a mod, that mod is a prime suspect for the crash. \n\n",
                "-----\n",
            ])
        return ReportFragment.empty()

File: classic-scanlog-core/src/suspect_scanner.rs
//! Suspect Scanner - Pattern matching for known crash suspects
//!
//! This module scans crash logs for known patterns and suspect errors:
//! - Main error checking against known patterns
//! - Call stack scanning for problematic signatures
//! - DLL-related crash detection
//! - YAML-defined suspect pattern matching

use crate::error::Result;
use std::collections::HashMap;
use rayon::prelude::*;

use crate::report::ReportFragment;

/// Signal modifiers for suspect detection
const MAIN_ERROR_REQUIRED: &str = "ME-REQ";
const MAIN_ERROR_OPTIONAL: &str = "ME-OPT";
const CALLSTACK_NEGATIVE: &str = "NOT";

/// Match status for suspect scanning
#[derive(Debug, Clone)]
struct MatchStatus {
    has_required_item: bool,
    error_req_found: bool,
    error_opt_found: bool,
    stack_found: bool,
}

impl MatchStatus {
    fn new() -> Self {
        Self {
            has_required_item: false,
            error_req_found: false,
            error_opt_found: false,
            stack_found: false,
        }
    }

    /// Check if this represents a suspect match
    fn is_suspect(&self) -> bool {
        if self.has_required_item {
            self.error_req_found
        } else {
            self.error_opt_found || self.stack_found
        }
    }
}

/// High-performance suspect scanner (40x speedup)
#[derive(Clone)]
pub struct SuspectScanner {
    suspects_error_list: HashMap<String, String>,
    suspects_stack_list: HashMap<String, Vec<String>>,
}

impl SuspectScanner {
    pub fn new(
        suspects_error_list: HashMap<String, String>,
        suspects_stack_list: HashMap<String, Vec<String>>,
    ) -> Self {
        Self {
            suspects_error_list,
            suspects_stack_list,
        }
    }

    /// Scan main error for suspect patterns
    ///
    /// Args:
    ///     crashlog_mainerror: The main error from crash log
    ///     max_warn_length: Maximum length for formatting error names
    ///
    /// Returns:
    ///     Tuple of (ReportFragment, bool indicating if suspects found)
    pub fn suspect_scan_mainerror(
        &self,
        crashlog_mainerror: &str,
        max_warn_length: usize,
    ) -> Result<(ReportFragment, bool)> {
        let mut lines = Vec::new();
        let mut found_suspect = false;

        for (error_key, signal) in &self.suspects_error_list {
            // Skip if signal not in crash log
            if !crashlog_mainerror.contains(signal.as_str()) {
                continue;
            }

            // Parse error information (format: "Severity | Error Name")
            if let Some((error_severity, error_name)) = error_key.split_once(" | ") {
                // Format error name for report
                let formatted_error_name = format!("{:.<width$}", error_name, width = max_warn_length);

                // Add to report
                lines.push(format!(
                    "- **Checking for {} SUSPECT FOUND! > Severity : {}** \n\n",
                    formatted_error_name, error_severity
                ));
                lines.push("-----\n".to_string());

                found_suspect = true;
            }
        }

        Ok((ReportFragment::from_lines(lines), found_suspect))
    }

    /// Analyze call stack for suspect patterns with signal modifiers
    ///
    /// Args:
    ///     crashlog_mainerror: Main error from crash log
    ///     segment_callstack_intact: Full call stack segment
    ///     max_warn_length: Maximum length for formatting error names
    ///
    /// Returns:
    ///     Tuple of (ReportFragment, bool indicating if suspects found)
    pub fn suspect_scan_stack(
        &self,
        crashlog_mainerror: &str,
        segment_callstack_intact: &str,
        max_warn_length: usize,
    ) -> Result<(ReportFragment, bool)> {
        let mut lines = Vec::new();
        let mut any_suspect_found = false;

        for (error_key, signal_list) in &self.suspects_stack_list {
            // Parse error information (format: "Severity | Error Name")
            let (error_severity, error_name) = match error_key.split_once(" | ") {
                Some(parts) => parts,
                None => continue,
            };

            // Track match status
            let mut match_status = MatchStatus::new();
            let mut should_skip = false;

            // Process each signal
            for signal in signal_list {
                if Self::process_signal(
                    signal,
                    crashlog_mainerror,
                    segment_callstack_intact,
                    &mut match_status,
                ) {
                    // NOT condition met, skip this error
                    should_skip = true;
                    break;
                }
            }

            if should_skip {
                continue;
            }

            // Check if we have a suspect match
            if match_status.is_suspect() {
                let formatted_error_name = format!("{:.<width$}", error_name, width = max_warn_length);
                lines.push(format!(
                    "- **Checking for {} SUSPECT FOUND! > Severity : {}** \n\n",
                    formatted_error_name, error_severity
                ));
                lines.push("-----\n".to_string());
                any_suspect_found = true;
            }
        }

        Ok((ReportFragment::from_lines(lines), any_suspect_found))
    }

    /// Check if crash involves a DLL file
    ///
    /// Args:
    ///     crashlog_mainerror: Main error from crash log
    ///
    /// Returns:
    ///     ReportFragment containing DLL crash notification, or empty
    pub fn check_dll_crash(crashlog_mainerror: &str) -> Result<ReportFragment> {
        let crashlog_lower = crashlog_mainerror.to_lowercase();

        if crashlog_lower.contains(".dll") && !crashlog_lower.contains("tbbmalloc") {
            Ok(ReportFragment::from_lines(vec![
                "* NOTICE : MAIN ERROR REPORTS THAT A DLL FILE WAS INVOLVED IN THIS CRASH! * \n".to_string(),
                "If that dll file belongs to a mod, that mod is a prime suspect for the crash. \n\n".to_string(),
                "-----\n".to_string(),
            ]))
        } else {
            Ok(ReportFragment::empty())
        }
    }
}

impl SuspectScanner {
    /// Process an individual signal and update match status
    ///
    /// Returns:
    ///     true if processing should stop (NOT condition met)
    fn process_signal(
        signal: &str,
        crashlog_mainerror: &str,
        segment_callstack_intact: &str,
        match_status: &mut MatchStatus,
    ) -> bool {
        // Check if signal has modifier
        if !signal.contains('|') {
            // Simple case: direct string match in callstack
            if segment_callstack_intact.contains(signal) {
                match_status.stack_found = true;
            }
            return false;
        }

        // Parse signal modifier and string
        let (signal_modifier, signal_string) = match signal.split_once('|') {
            Some(parts) => parts,
            None => return false,
        };

        // Process based on signal modifier
        match signal_modifier {
            MAIN_ERROR_REQUIRED => {
                match_status.has_required_item = true;
                if crashlog_mainerror.contains(signal_string) {
                    match_status.error_req_found = true;
                }
            }
            MAIN_ERROR_OPTIONAL => {
                if crashlog_mainerror.contains(signal_string) {
                    match_status.error_opt_found = true;
                }
            }
            CALLSTACK_NEGATIVE => {
                // Return true to break if NOT condition is met
                return segment_callstack_intact.contains(signal_string);
            }
            modifier if modifier.chars().all(|c| c.is_ascii_digit()) => {
                // Check for minimum occurrences
                if let Ok(min_occurrences) = modifier.parse::<usize>() {
                    let count = segment_callstack_intact.matches(signal_string).count();
                    if count >= min_occurrences {
                        match_status.stack_found = true;
                    }
                }
            }
            _ => {}
        }

        false
    }

    /// Batch process multiple crash logs for suspects (parallel)
    ///
    /// This provides significant speedup for batch operations
    pub fn scan_suspects_batch(
        &self,
        crash_logs: Vec<(String, String)>,  // Vec<(main_error, callstack)>
        max_warn_length: usize,
    ) -> Result<Vec<(ReportFragment, bool)>> {
        let results: Vec<_> = crash_logs
            .par_iter()
            .map(|(main_error, callstack)| {
                // Scan main error
                let (main_frag, main_found) = self
                    .suspect_scan_mainerror(main_error, max_warn_length)
                    .unwrap_or_else(|_| (ReportFragment::empty(), false));

                // Scan call stack
                let (stack_frag, stack_found) = self
                    .suspect_scan_stack(main_error, callstack, max_warn_length)
                    .unwrap_or_else(|_| (ReportFragment::empty(), false));

                // Combine results
                (main_frag.combine(&stack_frag), main_found || stack_found)
            })
            .collect();

        Ok(results)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_suspect_scan_mainerror() {
        let mut error_list = HashMap::new();
        error_list.insert(
            "Critical | Memory Access Violation".to_string(),
            "ACCESS_VIOLATION".to_string(),
        );

        let scanner = SuspectScanner::new(error_list, HashMap::new());

        let (fragment, found) = scanner
            .suspect_scan_mainerror("Error: ACCESS_VIOLATION at 0x12345", 50)
            .unwrap();

        assert!(found);
        assert!(fragment.len() > 0);
    }

    #[test]
    fn test_check_dll_crash() {
        let fragment = SuspectScanner::check_dll_crash(
            "Error in plugin.dll at address 0x12345"
        ).unwrap();

        assert!(fragment.len() > 0);

        // Should not trigger for tbbmalloc
        let fragment2 = SuspectScanner::check_dll_crash(
            "Error in tbbmalloc.dll"
        ).unwrap();

        assert!(fragment2.len() == 0);
    }

    #[test]
    fn test_signal_processing() {
        let mut match_status = MatchStatus::new();

        // Test ME-REQ signal
        SuspectScanner::process_signal(
            "ME-REQ|OutOfMemory",
            "OutOfMemory error occurred",
            "callstack here",
            &mut match_status,
        );

        assert!(match_status.has_required_item);
        assert!(match_status.error_req_found);
        assert!(match_status.is_suspect());
    }

    #[test]
    fn test_occurrence_count() {
        let mut match_status = MatchStatus::new();

        // Test occurrence count signal (3|pattern means pattern must appear 3+ times)
        SuspectScanner::process_signal(
            "3|SomePattern",
            "main error",
            "SomePattern\nSomePattern\nSomePattern\ndata",
            &mut match_status,
        );

        assert!(match_status.stack_found);
    }
}
